### ISSUE TITLE
How should language dialects be handled?

### ISSUE THREAD

**Issue by vermiculus** [google/code-prettify#370]
#115 introduced LaTeX syntax highlighting and was surely a boon to TeX.SX :smile:

However, it's come to light that TeX.SX deals also with a TeX dialect called `expl3`, a full-blown programming language built on top of TeX primitives.  It has quite a different syntax from LaTeX.  For my work in emacs, I created a syntax-highlighting scheme that recognizes `expl3` very well, so it's not difficult to sort through:
![scrsht](https://cloud.githubusercontent.com/assets/2082195/7381816/5e73e1be-edcd-11e4-8a11-79a0e2ba828b.png)
Notice how `_` and `:` are defined parts of the syntax now – something that's completely different from standard TeX (even though it still runs on TeX).

The only concern on my end (before working on and submitting a PR for `expl3`) is how exactly to *organize* any language dialect that significantly differs from the language it technically derives from.

**Comment by mikesamuel** [comment]
I wanted to piggyback on file extensions since people understand them even if they're not a proper partition.

I'd rather not add more knobs to the interface if we can avoid it.

That said, if there's a real need, we can figure something out.

The pattern used to find language hints is
https://github.com/google/code-prettify/blob/master/src/prettify.js#L1471
>     var langExtensionRe = /\blang(?:uage)?-([\w.]+)(?!\S)/;

so we could change the code that searches the extension registry to, given the language hint
   "lang-tex.expl3"
looks for the first key in
    "tex.expl3"
    "tex"
before falling back to its default heuristic.

**Comment by vermiculus** [comment]
Well, the need is real for me and for many of the regulars on TeX.SX :) having the familiar dot-extension seems like a good idea. 

Where would these language definitions go, though? The obvious answer is in a new file named after the key, but it could also be included in the master-language file. 

The former runs the risk of bloating the directory in time.

**Comment by mikesamuel** [comment]
> Well, the need is real for me and for many of the regulars on TeX.SX :)

I didn't mean to discount either.

But I'm not clear on the implications of 

> (even though it still runs on TeX)
 
If the regular tex mode could incorporate these syntactic differences without breaking on idiomatic snippets of non-expl3 tex then could tex+expl3 just be tex mode?

**Comment by vermiculus** [comment]
> tex+expl3 just be tex mode?

I'm not quite sure I understand the terminology, here.

**Comment by mikesamuel** [comment]
By "tex+expl3 just be tex mode" I just meant make the mode work for the dialect that you referenced.

**Comment by vermiculus** [comment]
Ah, but it can't :) For example,
```tex
\int_new:N \l_test_int
```
could mean 'integral subscript n times e time w…' just as easily as it means 'new int called l_test_int'. Perhaps more down-to-earth, consider
```tex
\hbox:n
```
could be either `:` in a box, then `n` (standard TeX) *or* a function to introduce content to be set in a box.

Syntax highlighting doesn't care how the language behaves, but it does care about what it considers a function, what it considers a type, etc.

**Comment by mikesamuel** [comment]
Fair enough.  If it's the case that most pages that use lang-tex do not use lang-tex.expl3 and vice versa then let's put them in separate js file.
